Automattic/wp-calypso
# Improve WpcomChecklist: reduce re-renders and load dynamically in stats

This PR optimizes performance around `WpcomChecklist` a bit by:

* Reducing re-renders in the `WpcomChecklist` component, caused by failing shallow comparisons in a prop coming from Redux. Applies memoization to the selector to fix the issue.
* Dynamically loading `WpcomChecklist` in stats, and only doing so when the site qualifies and the checklist hasn't been dismissed.

**Note to reviewers**: please pay special attention to the memoization code added in the first commit. I still haven't fully grokked the `createSelector` library, so I want to make sure I avoid introducing any subtle bugs. I couldn't find any, but, well, that's why they'd be subtle 😄 

Measured deltas in stats page, for sites without a checklist banner:
* ~12% improvement in the `perfmon` metric 👍 
* ~8KB less JS (gzipped) 👍 

Measured deltas in stats page, for sites with a checklist banner:
* No apparent performance improvement in the `perfmon` metric (< 5% change)
* ~500B more JS (gzipped) 👎 Seems acceptable, though.

**Testing instructions:**
* Select a site with no active checklist banner
* Open the stats page and ensure there are no issues

* Select a site with an active checklist banner
* Open the stats page and ensure there are no issues

**Performance analysis:**
* Create a build with `perfmon` enabled
* Set CPU throttling to 6x, to reduce sample noise
* Load the stats page once
* Alternate between the activity and stats pages, taking note of the `perfmon` values for the stats page. If possible, automate this, to reduce sample noise introduced by mouse pointer movement.

Perform the process above for a site with a checklist banner, and one without.
